lxd/storage/drivers/driver_pure_util.go

```diff
@@ -517,6 +572,475 @@ func (p *pureClient) disconnectHostFromVolume(poolName string, volName string, h
+		if err != nil {
+			return nil, fmt.Errorf("Failed to connect to PureStorage array %q via iSCSI: %s\n%v", targetAddr, stderr, err)
+		}
+	}
+
+	revert.Add(func() { _ = d.disconnect(*target) })
+
+	cleanup := revert.Clone().Fail
+	revert.Success()
+	return cleanup, nil
+}
+
+// disconnect disconnects this host from the given target with a given qualified name.
+func (d *pure) disconnect(target pureTarget) error {
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		targetIQN := *target.IQN
+
+		sessionID, err := iscsiFindSession(targetIQN)
+		if err != nil {
+			return err
+		}
+
+		if sessionID != "" {
+			// Disconnect from the iSCSI target.
+			_, err := shared.RunCommand("iscsiadm", "--mode", "node", "--targetname", targetIQN, "--logout")
+			if err != nil {
+				return fmt.Errorf("Failed disconnecting from PureStorage iSCSI target %q: %w", targetIQN, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// mapVolume maps the given volume onto this host.
+func (d *pure) mapVolume(vol Volume) error {
+	revert := revert.New()
+	defer revert.Fail()
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return err
+	}
+
+	unlock, err := locking.Lock(d.state.ShutdownCtx, d.config["pure.mode"])
+	if err != nil {
+		return err
+	}
+
+	defer unlock()
+
+	// Ensure the host exists and is configured with the correct QN.
+	hostname, cleanup, err := d.ensureHost()
+	if err != nil {
+		return err
+	}
+
+	revert.Add(cleanup)
+
+	// Ensure the volume is connected to the host.
+	connCreated, err := d.client().connectHostToVolume(vol.pool, volName, hostname)
+	if err != nil {
+		return err
+	}
+
+	if connCreated {
+		revert.Add(func() { _ = d.client().disconnectHostFromVolume(vol.pool, volName, hostname) })
+	}
+
+	// Connect to the array using iscsi. Connection can be established only
+	// when at least one volume is connected to the host.
+	cleanup, err = d.connect()
+	if err != nil {
+		return err
+	}
+
+	revert.Add(cleanup)
+
+	revert.Success()
+	return nil
+}
+
+// unmapVolume unmaps the given volume from this host.
+func (d *pure) unmapVolume(vol Volume) error {
+	var host *pureHost
+	var iqn string
+	var err error
+
+	volName, err := d.getVolumeName(vol)
+	if err != nil {
+		return err
+	}
+
+	unlock, err := locking.Lock(d.state.ShutdownCtx, d.config["pure.mode"])
+	if err != nil {
+		return err
+	}
+
+	defer unlock()
+
+	switch d.config["pure.mode"] {
+	case pureModeISCSI:
+		iqn, err = d.hostIQN()
+		if err != nil {
+			return err
+		}
+
+		host, err = d.client().getHostByIQN(iqn)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Disconnect the volume from the host and ignore error if connection does not exist.
+	err = d.client().disconnectHostFromVolume(vol.pool, volName, host.Name)
+	if err != nil && !api.StatusErrorCheck(err, http.StatusNotFound) {
+		return err
+	}
+
+	// Wait until the volume has disappeared.
```

Comment:
this comment seems to be in the wrong place, as it should be down by waitGone, as we've not yet request it be disconnected by this point, so its not going to have disappeared yet right?